### DISCUSSION TITLE
Where is the spindle - how to find in gcode program

### DISCUSSION THREAD

**Discussion by pad73** [terjeio/ioSender#333] [General]
Does exist way how to find on what line of gcode program is when milling? Or knows ioSender to show on monitor (3D data view) with some cursor or colour point exact position of selected gcode line?
For example in Candle sender you can list in code and small colour point is running on gcode program lines. It is easy to find some position on monitor (viewed data) listing in program. Or another example, in linuxCNC you can click with cursor to displayed data of program, and i will show you line where these data are.
Sometimes you need to continue program not from start but from place you need (for example after broking of the mill bit). I don´t know how to find this position.

**Comment by phil-barrett** [comment]
Terje can do a much better job of explaining ioSender than I.

I believe that Candle is only telling you the current line being sent to Grbl, not the one currently being executed.  You can test that by constructing GCode that has a very slow, long move followed by a number of other moves.  I think you will see candle sending the subsequent lines before the completion of the log, slow one.

On the issue of resuming a program after an interruption.  It isn't quite so simple as starting at the last good line.  I generally try to restart at the most recent move where speed was set (I.E. a block with an Sxxxx command).  Usually, it is not a large amount of GCode between those points. Restarting is a complex issue in general and any software that claims to do it probably has a number of caveats.

**Reply by pad73** [reply]
You are true and you aren´t. Candle is very basic user friendly software for sending data. I believe most of users dont need other program. It shows everything important. And knows making heightmap very easily.
About current spindle position. You can see on graphic two points. One shows what data are "in the memory" and second colour point show actual position. If you have very slow PC in combination with complicated gcode you can see these points are with very small distance between. In extreme situation milling is interrupted because chip has no data to execute.
About showing. I like when I can see both informations together - visual control of program and gcode lines. But you can see only one panel - Program / 3D View.
![image](https://github.com/terjeio/ioSender/assets/31275365/2ac74bff-bab2-4857-abb1-b6813614870f)

And of course, when program is going you can see executed lines in gcode, there is OK written, but when you will stop the program "OK" symbols are lost, so you dont know where program was stopped. When you pause the program using Feed Hold "OK" will stay. But it isn´t problem, it is understable. Problem is you are not able to find what gcode line is on the graphic panel. And yes, Terjeio on some other thread wrote that it isnt good idea to start exactly from last executed line, but you need to find optional line (where coordinates XYZ are precisely defined, where speed is defined, RPM ... etc). But, Im repeat the same, you need to find gcode line where you stopped.

**Reply by terjeio** [reply]
The Candle version I have does not even display a graphic representation of the gcode... Which version do you have that can also create and apply a heightmap?

> ... but when you will stop the program "OK" symbols are lost

Try the [latest edge version](http://www.io-engineering.com/downloads), it does not clear the `OK` and `*` block marks on stop. `*` means the block has been sent to the controller but not yet been acknowledged.

Adding code for restarting from a certain block is something that can be added but needs a carefully crafted specification and rigorous testing. Testing would be the most time consuming part of it...

**Reply by pad73** [reply]
This is picture from graphic representation. You can see violet line with violet point at the end. Point show data loaded. Orange pencil is spindle symbol.
![image](https://github.com/terjeio/ioSender/assets/31275365/71f364fa-5cc4-4e33-a884-6a79ebf1ad81)
I made simple manual for Candle for my friends, czech language only https://lupa-cnc.com/grbl-candle/

leaving OK on processed lines works well, thank you. But now I found .... may I start not from start in ioSender?

Now I need only one thing .... XL version with these addons :)

And for end, Im sorry, I dont want start new thread so I will use this. 
![image](https://github.com/terjeio/ioSender/assets/31275365/71b64a19-5f93-4cb8-a44a-5421c397fddf)
Why text and +- buttons are not in line. Or better describtion - why Rapid text is between +- for normal moving (by program) and +- for rapid moving?

**Reply by terjeio** [reply]
> But now I found .... may I start not from start in ioSender?

I can relatively easily add a right-click menu in the program listing with a _Start from here_ option, but with the caveat that you have to put the controller in the correct state before doing so. Adding code to ensure that the controller is in the correct state is a somewhat larger task.

> Now I need only one thing .... XL version with these addons :)

The graphics rendering library I use does not make this easy, I have already been experimenting with rendering the executing line in a different color, this can be enabled by editing the _App.config_ file but I do not recommend doing so as it affected performance negatively - especially so for laser engravings. Back and forth highlighting from _Program_ <> _3D View_ might be possible but would require a new layout of the senders in addition to supporting code? I'll put it on my todo list.

> why Rapid text is between +- for normal moving (by program) and +- for rapid moving?

Because normal motion can be overridden in both small and big steps in both directions (+ and -), rapids only in the negative direction in two large steps. 0 is for resetting to 100%.

**Reply by phil-barrett** [reply]
> I can relatively easily add a right-click menu in the program listing with a Start from here option, but with the caveat that you have to put the controller in the correct state before doing so. Adding code to ensure that the controller is in the correct state is a somewhat larger task.

If you do that, It would be great if you could also add a way for the user to identify the prelude/setup blocks at the beginning of the GCode program.  It is usually fairly easy to recognize though each GCode generating program is different so probably not easily automatable.  Then, with the prelude identified, restart execution by executing the prelude and then skip to the restart block.  I would be ok with writing out a new restart file to be executed from.

I'm not sure how this would work with tool change and other things though.

**Reply by terjeio** [reply]
> If you do that, It would be great if you could also add a way for the user to identify the prelude/setup blocks at the beginning of the GCode program.

Not as a part of the first step towards restarting from a selected position as this will complicate the code and require a lot more testing. Right click on a block, select "start from here", acknowledge a warning/message asking "Do you know what you are about to do now?" is likely easy for me and leaves the responsibility 100% with the user if something goes wrong.

If the program is loaded it is easy to open it in an editor (from the toolbar) and copy/paste the blocks/lines into a new file and from there to the MDI. Or perhaps I should add "Paste to the MDI" as a right click option? And even "Send to controller" as well?

**Comment by pad73** [comment]
> > But now I found .... may I start not from start in ioSender?
> 
> I can relatively easily add a right-click menu in the program listing with a _Start from here_ option, but with the caveat that you have to put the controller in the correct state before doing so. Adding code to ensure that the controller is in the correct state is a somewhat larger task.

So if possible please add this to the ioSender

**Comment by terjeio** [comment]
New [edge version](http://www.io-engineering.com/downloads) uploaded with right click popup menu in the program listing with items for starting from the selected block, copy the command to the MDI or sending one or more commands to the controller. The last sorts the commands in ascending order regardless of how the commands were selected. I have not extensively tested this, be careful when trying it out and report any unexpected behaviour...

Tip: turn on _Parser state_ in the _$10 Status report options_, this will display the current controller state:

![image](https://github.com/terjeio/ioSender/assets/20260062/8895b04e-b598-4df2-a9a6-208779f9901b)

This might be of help when ensuring the state is ok before restarting a program.

**Reply by pad73** [reply]
I tested this last version 2.0.44.p6 (XL)
Start program from selected line works fine. 
But ...

Turning on $10 Parser state makes nothing 
![image](https://github.com/terjeio/ioSender/assets/31275365/d979408d-574d-403c-a459-e5c63df46952)
![image](https://github.com/terjeio/ioSender/assets/31275365/668961df-808e-40be-a17d-efaaecdc2c6f)


Again, the key controls are lost, the shortcuts don't work in this version.

**Reply by terjeio** [reply]
> Turning on $10 Parser state makes nothing

Did you try restarting the sender after enabling _Parser state_?

> Again, the key controls are lost, the shortcuts don't work in this version.

It is the XL version that does not work with some shortcuts, and has not for a long time? New edge version with fix uploaded.

**Reply by pad73** [reply]
Yes, I restarted the iosender

**Comment by pad73** [comment]
Hallo after a quite long time. I can see new version .44 is here. Looks like Shortcut issue solved (Alt+R to start milling etc.). But "highlight completed cuts" is still impossible to activate. Is it only my problem (depended, may be, on used board), or is it "standard"? It is interesting that this function is working with Candle sender.
